src/Prototype/Types.hs

```diff
@@ -114,16 +116,20 @@ instance FromJSON Session
 -- The signed data we write to/read from cookies. Should not be used outside
 -- the authentication layer.
 data User = User
-  { username :: Namespace
-  , email    :: Text
+  { username    :: Namespace -- ^ User's unique namespace/id
+  , email       :: Text -- ^ TODO: newtype 
+  , userGroups  :: Set GroupId -- ^ Set of groups the user belongs to 
```

Comment:
Ah, I'm not entirely sure this should be added here, since `User` is the data that is put into cookies, and that I think we want to lookup a real `Profile` anyway. It's a bit like we shove more that required in the cookie (and since it is a prototype, I feel like the cookie should only be there for authentication, as its sole purpose). What do you think ?

Reply:
I personally feel that putting the user into a cookie is a little problemmatic. The idea is to rely on a cookie, and the cookie alone for user authentication (esp. if the cookie has been signed and the signature can be verified); but this gives us the issue you're pointing out: the state of the user shouldn't be kept in the cookie at all. 
I think we should only keep information specific to the authentication (as JWT) in the cookie; and look up the user in our database for each verification. 

So, I'd put the following in the cookie:
1. User's ID (namespace) to uniquely identify the user 
2. Time of login 
3. Time of validity of the login 
4. Ideally a unique session-id. 

Now, when the user logs in successfully, this information can be generated and stored in the cookie. 
For a logged-in user, we look up this cookie, verify its signature, and then start verifying its contents. 

1. Look up the user with the user-id, if the user exists, well and good, we now have the user-information in our hands.
2. Verify the session-id in the list of ongoing user-sessions (for this user)
3. If the user is attempting to access a resource, verify the permissions needed. 

When the user logs out, we invalidate the session-id, and be done with it. 

---

The issue with storing user information in cookies exposes too much logic to the readers of the cookie. This can be overcome by fully encrypting the cookie as well. Also, if we start relying on the information in the cookie; we have situations like these: not just for group ids but for other user information as well.

Eg. what if the user also has a firstname field, and they want to change that. If we read this information from the cookie alone, we run into an issue. The updated user information will not show up in cookies unless we rewrite the cookie. 

Thoughts?

Reply:
I think we were saying the same thing : the username is enough. As for the time of logging and expiration, I guess this would be required for a "real" application, not for a prototype, but why not. Finally, if we have a session ID, and I guess it makes sense, and we always want to look it up together with the user, maybe the username becomes redundant.